actionview/lib/action_view/helpers/form_helper.rb

```diff
@@ -2722,7 +2722,13 @@ def fields_for_with_nested_attributes(association_name, association, options, bl
             output = ActiveSupport::SafeBuffer.new
             association.each do |child|
               if explicit_child_index
-                options[:child_index] = explicit_child_index.call if explicit_child_index.respond_to?(:call)
+                if explicit_child_index.respond_to?(:call)
+                  if explicit_child_index.arity.zero?
+                    options[:child_index] = explicit_child_index.call
+                  else
+                    options[:child_index] = explicit_child_index.call(child)
+                  end
+                end
```

Comment:
Could this set of branches be flattened based on the arity?

```suggestion
                if explicit_child_index.respond_to?(:call)
                  arguments = [child].take(explicit_child_index.arity)
                  options[:child_index] = explicit_child_index.call(*arguments)
                end
```

Reply:
Or, if that's too clever:

```suggestion
                if explicit_child_index.respond_to?(:call)
                  options[:child_index] =
                    if explicit_child_index.arity.zero?
                      explicit_child_index.call
                    else
                      explicit_child_index.call(child)
                    end
                end
```

Reply:
Your first suggestion make sense, but to be honest I did have to verify the following in `irb`

```ruby
> child_index = -> { "hi" }
> child_index.call(*[])
=> "hi"
```